cache/cache_reservation_manager.h

```diff
@@ -172,20 +213,61 @@ class CacheReservationManager
   CacheKey cache_key_;
 };
 
-// CacheReservationHandle is for managing the lifetime of a cache reservation
-// This class is NOT thread-safe
-template <CacheEntryRole R>
-class CacheReservationHandle {
+class ConcurrentCacheReservationManager
+    : public CacheReservationManager,
+      public std::enable_shared_from_this<ConcurrentCacheReservationManager> {
  public:
-  // REQUIRES: cache_res_mgr != nullptr
-  explicit CacheReservationHandle(
+  ConcurrentCacheReservationManager(
+      std::shared_ptr<CacheReservationManager> cache_res_mgr) {
+    cache_res_mgr_ = cache_res_mgr;
+  }
+  ConcurrentCacheReservationManager(const ConcurrentCacheReservationManager &) =
+      delete;
+  ConcurrentCacheReservationManager &operator=(
+      const ConcurrentCacheReservationManager &) = delete;
+  ConcurrentCacheReservationManager(ConcurrentCacheReservationManager &&) =
+      delete;
+  ConcurrentCacheReservationManager &operator=(
+      ConcurrentCacheReservationManager &&) = delete;
+
+  ~ConcurrentCacheReservationManager(){};
+
+  inline Status UpdateCacheReservation(std::size_t new_memory_used) override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    return cache_res_mgr_->UpdateCacheReservation(new_memory_used);
+  }
+  inline Status MakeCacheReservation(
       std::size_t incremental_memory_used,
-      std::shared_ptr<CacheReservationManager> cache_res_mgr);
+      std::unique_ptr<CacheReservationManager::CacheReservationHandle> *handle)
+      override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    Status s = cache_res_mgr_->UpdateCacheReservation(
+        cache_res_mgr_->GetTotalMemoryUsed() + incremental_memory_used);
+    (*handle).reset(new CacheReservationHandle(incremental_memory_used,
+                                               shared_from_this()));
 
-  ~CacheReservationHandle();
+    return s;
+  }
+  inline std::size_t GetTotalReservedCacheSize() override {
+    return cache_res_mgr_->GetTotalReservedCacheSize();
+  }
+  inline std::size_t GetTotalMemoryUsed() override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    return cache_res_mgr_->GetTotalMemoryUsed();
+  }
 
  private:
-  std::size_t incremental_memory_used_;
+  std::mutex cache_res_mgr_mu_;
   std::shared_ptr<CacheReservationManager> cache_res_mgr_;
+
+  inline Status ReleaseCacheReservation(
+      std::size_t incremental_memory_used) override {
+    std::lock_guard<std::mutex> lock(cache_res_mgr_mu_);
+    assert(cache_res_mgr_->GetTotalMemoryUsed() >= incremental_memory_used);
+    std::size_t updated_total_mem_used =
+        cache_res_mgr_->GetTotalMemoryUsed() - incremental_memory_used;
+    Status s = cache_res_mgr_->UpdateCacheReservation(updated_total_mem_used);
+    return s;
+  }
```

Comment:
Hm when complaining "there are two CacheReservationHandle classes" what I really meant to complain is there are two unrelated handle classes. But the handle destructor is the public interface to releasing memory, so it makes sense to have multiple. A ConcurrentCacheReservationManager::CacheReservationHandle could wrap a CacheReservationManagerImpl::CacheReservationHandle and provide locking.

Reply:
btw I tried out this suggestion here in case it's useful - https://github.com/ajkr/rocksdb/commit/87f73be5acb0158ea0c8483dc1497528550d39ca

Reply:
> btw I tried out this suggestion here in case it's useful - ajkr@87f73be

Make sure this is visible since the "Add single comment" appears to not make comments visible in the same way as a full review (?).

Reply:
> btw I tried out this suggestion here in case it's useful - ajkr/rocksdb@87f73be

Make sure this is visible since the "Add single comment" appears to not make comments visible in the same way as a full review (?).

Reply:
Done - similar to your suggestion excepts that (a) no template for  CacheReservationManagerImpl<R>::CacheReservationHandle (b) remove useless access method for std::size_t incremental_memory_used in (c) make CacheReservationManager::CacheReservationHandle more as an interface (i.e, no data stored) to be extended (just like CacheReservationManager)

Reply:
> (a) no template for CacheReservationManagerImpl::CacheReservationHandle

Good finding, I did not know C++ does this.